### PR TITLE
Base implementation of SVGP

### PR DESCRIPTION
**Pull request by rossviljoen**
This PR has the core implementation of stochastic variational GPs.

The approximate posterior in `svgp.jl` quite closely follows that in AbstractGPs. `elbo.jl` has the implementations of both the closed form Gaussian likelihood ELBO and quadrature for non-conjugate - I'll add other closed form likelihoods and MC estimation in other PRs.

The examples in this PR are extremely rough, and will be superseded soon, but give working examples of how to train the SVGP end to end.

### CODE REVIEW

File: src/svgp.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+q(f^*) = \\int p(f | u) q(u) du
+```
+which can be found in closed form.
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+
+function approx_posterior(::SVGP, fz::FiniteGP, q::MvNormal)
```
Comment (theogf) [inline src/svgp.jl:21]: Instead of `MvNormal` you could expect and `AbstractMvNormal` and instead of `q.μ` call `mean(q)`

File: examples/classification.jl [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+# %%
+# After optimisation, plot samples from the underlying posterior GP.
+
+fu = f(z).fx # want the underlying FiniteGP
+post = SparseGPs.approx_posterior(SVGP(), fu, MvNormal(m, A'A))
+l_post = LatentGP(post, BernoulliLikelihood(), 0.1)
+
+post_f_samples = rand(l_post.f(x_plot, 1e-6), 20)
+
```
Comment (theogf) [inline examples/classification.jl:136]: It would be nice to plot the mean as well

**Review by theogf — COMMENTED** [review]
Main comment is that we should generally accept `q` as an `AbstractMvNormal`

File: examples/classification.jl (hunk 2 above)
Comment (rossviljoen) [reply examples/classification.jl:136]: I'm going to rework the examples a lot, so I won't add that in this PR but will do later

**Comment by rossviljoen** [comment]
Thanks for the review! I changed everything to use `AbstractMvNormal`

File: src/elbo.jl [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    y::AbstractVector,
+    f_mean::AbstractVector,
+    f_var::AbstractVector,
+    lik::ScalarLikelihood;
+    n_points=20
+)
+    return sum(gauss_hermite_quadrature(y, f_mean, f_var, lik; n_points=n_points))
+end
+
+function kl_divergence(q::AbstractMvNormal, p::AbstractMvNormal)
```
Comment (st--) [inline src/elbo.jl:124]:
```suggestion
function StatsBase.kldivergence(q::AbstractMvNormal, p::AbstractMvNormal)
```
(and changing all calls from `kl_divergence` to `kldivergence`, of course)

I was expecting to see implementations from Distributions.jl already, but all I could find was https://github.com/theogf/KLDivergences.jl - would it be worth building on that @theogf ?

File: src/elbo.jl [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    # using a reparameterisation by change of variable
+    # (see eg. en.wikipedia.org/wiki/Gauss%E2%80%93Hermite_quadrature)
+    xs, ws = gausshermite(n_points)
+    # size(fs): (length(y), n_points)
+    fs = √2 * .√f_var .* transpose(xs) .+ f_mean
+    lls = loglikelihood.(lik.(fs), y)
+    return (1/√π) * lls * ws
+end
+
+ChainRulesCore.@non_differentiable gausshermite(n)
```
Comment (st--) [inline src/elbo.jl:148]: what makes it non-differentiable ? it's just a dot product, which should be easy to differentiate?

File: src/svgp.jl [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+Compute the approximate posterior [1] over the process `f = fz.f`, given inducing
+inputs `z = fz.x` and a variational distribution over inducing points `q(u)` where `u =
+f(z)`. The approximate posterior at test points ``x^*`` where ``f^* = f(x^*)``
+is then given by:
+
+```math
+q(f^*) = \\int p(f | u) q(u) du
+```
+which can be found in closed form.
```
Comment (st--) [inline src/svgp.jl:14]:
``````suggestion
raw"""
    approx_posterior(::SVGP, fz::FiniteGP, q::MvNormal)

Compute the approximate posterior [1] over the process `f = fz.f`, given inducing
inputs `z = fz.x` and a variational distribution over inducing points `q(u)` where `u =
f(z)`. The approximate posterior at test points ``x^*`` where ``f^* = f(x^*)``
is then given by:

```math
q(f^*) = \int p(f | u) q(u) du
```
which can be found in closed form.
``````
if you're not using string interpolation (i.e. basically all docstrings, the `raw` prefix makes it a lot easier to write (and read) LaTeX formulas:)

**Review by st-- — COMMENTED** [review]
Pass 1: elbo.jl and svgp.jl at high level - haven't checked math implementations in detail yet.

File: test/svgp.jl [hunk 6]
```diff
@@ -0,0 +1,4 @@
+@testset "svgp" begin
```
Comment (st--) [inline test/svgp.jl:1]: Could be removed now that there's actual tests being run as well

File: src/elbo.jl (hunk 3 above)
Comment (theogf) [reply src/elbo.jl:124]: Well I started building this to take it out of AGP.jl, but it is quite "brittle"

File: src/elbo.jl [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+"""
+
+function expected_loglik end
+
+"""
+    expected_loglik(y::AbstractVector{<:Real}, f_mean::AbstractVector, f_var::AbstractVector, Σy::AbstractVector)
+
+The expected log likelihood for a Gaussian likelihood, computed in closed form.
+"""
+# The closed form expected loglikelihood for a Gaussian likelihood
```
Comment (st--) [inline src/elbo.jl:97]:
```suggestion
```
already covered by the docstring:)

File: src/elbo.jl (hunk 3 above)
Comment (rossviljoen) [reply src/elbo.jl:121]: Ah, I didn't see that package - I suppose I'll keep the implementation here for now, and replace it if it gets more stable. @theogf what makes it brittle atm? I'd be up for working on it if needed.

File: src/elbo.jl (hunk 4 above)
Comment (rossviljoen) [reply src/elbo.jl:145]: `gausshermite(n)` just returns the nodes and weights, so it's a constant for a given n - it also does a load of mutation so it's not differentiable anyway

File: test/svgp.jl (hunk 6 above)
Comment (rossviljoen) [reply test/svgp.jl:1]: Yes, sorry that's still WIP - I need to look through https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/pull/179 and see what I can use from that here

File: src/elbo.jl (hunk 3 above)
Comment (theogf) [reply src/elbo.jl]: > what makes it brittle atm

I don't think it's very well tested and does not account for edge cases...

Let me have a look back to what I did, for example I did not know StatsBase.kldivergence existed, I should probably reuse that!

File: test/svgp.jl (hunk 6 above)
Comment (rossviljoen) [reply test/svgp.jl:1]: I've added the standard AbstractGP interface tests

File: src/elbo.jl [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    fx::FiniteGP,
+    y::AbstractVector{<:Real},
+    fz::FiniteGP,
+    q::AbstractMvNormal;
+    n_data=length(y)
+)
+    n_batch = length(y)
+    kl_term, f_mean, f_var = _elbo_intermediates(fx, fz, q)
+
+    Σy = diag(fx.Σy) # n.b. this assumes uncorrelated observation noise
```
[the remark this answers is not in the document]
Comment (rossviljoen) [reply src/elbo.jl]: I've refactored this quite a lot - doing something similar to what you suggested - so hopefully it's a bit nicer now?

File: src/elbo.jl [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+`q(f)` is an approximation to the latent function values `f` given by:
+```math
+    q(f) = ∫ p(f | u) q(u) du
+```
+where `q(u)` is the variational distribution over inducing points (see
+[`elbo`](@ref)).
+
+Where possible, this expectation is calculated in closed form. Otherwise, it is
+approximated using Gauss-Hermite quadrature by default.
```
[the remark this answers is not in the document]
Comment (rossviljoen) [reply src/elbo.jl:91]: I've added Monte Carlo as well as a way of picking a method (via the `method` argument to `elbo`). 

btw I'm using symbols for the argument, but I'm not sure if that's the most idiomatic way - should I use strings/enums instead?

File: src/elbo.jl [hunk 10]
```diff
@@ -1,6 +1,21 @@
 "Likelihoods which take a scalar (or vector of scalars) as input and return a single scalar."
 ScalarLikelihood = Union{BernoulliLikelihood,PoissonLikelihood,GaussianLikelihood}
 
+
+abstract type ExpectationMethod end
+struct Default <: ExpectationMethod end
+struct Analytic <: ExpectationMethod end
+
+struct GaussHermite <: ExpectationMethod
+    n_points
```
Comment (theogf) [inline src/elbo.jl:10]:
```suggestion
    n_points::Int
```

Here you can enforce the type

File: src/elbo.jl [hunk 11]
```diff
@@ -0,0 +1,235 @@
+"Likelihoods which take a scalar (or vector of scalars) as input and return a single scalar."
+ScalarLikelihood = Union{BernoulliLikelihood,PoissonLikelihood,GaussianLikelihood}
+
+
+abstract type ExpectationMethod end
+struct Default <: ExpectationMethod end
+struct Analytic <: ExpectationMethod end
+
+struct GaussHermite <: ExpectationMethod
```
Comment (theogf) [inline src/elbo.jl:9]: Maybe `Quadrature` would be more adapted, especially if we want to consider different approaches than `GaussHermite` later

Comment (rossviljoen) [reply src/elbo.jl:9]: That makes sense - also, do you think I should call them `DefaultExpectation`, `AnalyticExpectation` etc... instead or leave them as they are?

Comment (willtebbutt) [reply src/elbo.jl:9]: Personally, I think this is fine for now. Lets just get this merged. We can break / generalise it later if needed.

**Comment by rossviljoen** [comment]
@st-- @willtebbutt @theogf 
I've updated the variational expectation stuff to dispatch on types instead.

One annoying thing is that there's a lot of special casing with `expected_loglik` for the conjugate GP case (i.e. when using a `FiniteGP` instead of a `LatentFiniteGP`, so there's a `Σy` instead of a `lik`). I suppose it might be possible to get rid of this by constructing a `HeteroskedasticGaussianLikelihood`, but that assumes that both `mu` and `sigma` are functions of the latent process, so that might end up being pretty awkward to use anyway.

File: src/elbo.jl [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+the ELBO. The options are: `Default()`, `Analytic()`, `Quadrature()` and
+`MonteCarlo()`. For likelihoods with an analytic solution, `Default()` uses this
+exact solution. If there is no such solution, `Default()` either uses
+`Quadrature()` or `MonteCarlo()`, depending on the likelihood.
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+function elbo(
```
Comment (theogf) [inline src/elbo.jl:36]: To answer your question what is assumed to be computed here if we have no information about the likelihood?

Comment (theogf) [reply src/elbo.jl:36]: If I understand correctly, we implicitly use the Gaussian likelihood, so one could simply call `elbo(LatentFiniteGP(fx, GaussianLikelihood()), y, fz, q; ...)` or something similar

Comment (rossviljoen) [reply src/elbo.jl:36]: Yes, exactly, but what if `fx.Σy` is not homoscedastic - I can't figure out a good way to do something like `elbo(LatentFiniteGP(fx, GaussianLikelihood(diag(fx.Σy)), y, fz, q; ...)` if that makes sense

Comment (rossviljoen) [reply src/elbo.jl:36]: Unless we just assume it's homoscedastic for now and use `GaussianLikelihood(Σy[1])` as I already do for MC/quadrature for the Gaussian likelihood

Comment (willtebbutt) [reply src/elbo.jl:36]: Could definitely do that -- would make sense to me.

Might be an idea to dispatch on `fx.Σy` being a `Diagonal{<:Real, <:Fill}` or something though to prevent silent failure.

Comment (theogf) [reply src/elbo.jl:36]: Using heteroskedastic likelihoods is definitely not standard, I think we can ignore this case as a default.

Comment (rossviljoen) [reply src/elbo.jl:36]: Alright, I'll do that. @willtebbutt should I maybe just throw a warning if the matrix isn't explicitly `Diagonal`? otherwise it seems a bit overzealous to only dispatch on that?

Comment (willtebbutt) [reply src/elbo.jl:36]: Good idea. You could also add a method which throws an informative error if you get something other than a `Diagonal{<:Real, <:Fill}`.

Comment (st--) [reply src/elbo.jl:39]: > Using heteroskedastic likelihoods is definitely not standard, I think we can ignore this case as a default.

Allowing for heteroskedastic likelihoods is actually very relevant in practice, it would be really good to keep that use-case in mind in the design instead of inadvertently making it really hard. The key difference to "scalar" likelihoods is that now there are several latent outputs for each observation.

So this requires multi-output GPs - but for example with mixing kernels the number of "latent outputs" (what the likelihood sees) can be different from the number of "latent GPs" (what we parametrize with q(u))...

File: src/elbo.jl [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+variational distribution over inducing points `u = f(z)`.
+
+`method` selects which method is used to calculate the expected loglikelihood in
+the ELBO. The options are: `Default()`, `Analytic()`, `Quadrature()` and
+`MonteCarlo()`. For likelihoods with an analytic solution, `Default()` uses this
+exact solution. If there is no such solution, `Default()` either uses
+`Quadrature()` or `MonteCarlo()`, depending on the likelihood.
+
+N.B. the observation noise `fx.Σy` is assumed to be homoscedastic and
+uncorrelated - i.e. only `fx.Σy[1]` is used.
```
Comment (st--) [inline src/elbo.jl:33]: I'm confused by what `fx.Σy[1]` refers to. I thought `Σy` was a matrix (potentially diagonal/scaled identity) ?

File: src/elbo.jl [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+`q(f)` is an approximation to the latent function values `f` given by:
+```math
+    q(f) = ∫ p(f | u) q(u) du
+```
+where `q(u)` is the variational distribution over inducing points (see
+[`elbo`](@ref)). The marginal means and variances of `q(f)` are given by
+`f_mean` and `f_var` respectively.
+
+Where possible, this expectation is calculated in closed form. Otherwise, it is
+approximated using either Gauss-Hermite quadrature or Monte Carlo.
```
Comment (st--) [inline src/elbo.jl:107]: should this be updated to clarify that `method` determines how the expectation is evaluated?

File: src/elbo.jl [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+The expected log likelihood for a `ScalarLikelihood`, computed via `method`.
+Defaults to a closed form solution if it exists, otherwise defaults to
+Gauss-Hermite quadrature.
+"""
+function expected_loglik(
+    ::Default,
+    y::AbstractVector,
+    f_mean::AbstractVector,
+    f_var::AbstractVector,
```
Comment (st--) [inline src/elbo.jl:127]: Just wondering aloud: would there be any (significant) runtime cost (overhead) in representing this as an `AbstractVector{<:Normal}` instead?

File: src/svgp.jl [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+```math
+q(f^*) = \int p(f | u) q(u) du
+```
+which can be found in closed form.
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+
```
Comment (st--) [inline src/svgp.jl:20]:
```suggestion
```

File: examples/regression.jl [hunk 17]
```diff
@@ -0,0 +1,178 @@
+# A recreation of https://gpflow.readthedocs.io/en/master/notebooks/advanced/gps_for_big_data.html
```
Comment (st--) [inline examples/regression.jl:1]: @rossviljoen what are the steps to render the notebook? I tried `Literate.notebook("examples/regression.jl"; execute=true)`, but the resulting .ipynb doesn't actually contain any plots. (I also had to add a couple dependencies - Optim and IterTools are used by the notebook but not in the SparseGPs.jl/Project.toml)

File: src/elbo.jl [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+    lik::ScalarLikelihood
+)
+    method = _default_method(lik)
+    expected_loglik(method, y, f_mean, f_var, lik)
+end
+
+# The closed form solution for independent Gaussian noise
+function expected_loglik(
+    ::Analytic,
+    y::AbstractVector,
```
Comment (st--) [inline src/elbo.jl:143]: This likelihood requires `y` to be a number, so should this be
```suggestion
    y::AbstractVector{<:Real},
```
?

(Side note: My understanding is that the way KernelFunctions and AbstractGPs are designed, the elements of `y` are all scalars, as multi-output handling is based around the output-as-input formulation. How would we actually handle homotopic multi-output efficiently? probably one for @willtebbutt)

The same question also applies to the methods below.

File: src/elbo.jl [hunk 19]
```diff
@@ -0,0 +1,244 @@
+"Likelihoods which take a scalar (or vector of scalars) as input and return a single scalar."
```
Comment (st--) [inline src/elbo.jl:1]: what would be the "vector of scalars" case ?

File: src/elbo.jl [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+    return sum(-0.5 * (log(2π) .+ log.(lik.σ²) .+ ((y .- f_mean).^2 .+ f_var) / lik.σ²))
+end
+
+# The closed form solution for a Poisson likelihood with an exponential inverse link function
+function expected_loglik(
+    ::Analytic,
+    y::AbstractVector,
+    f_mean::AbstractVector,
+    f_var::AbstractVector,
+    ::PoissonLikelihood{ExpLink}
```
Comment (st--) [inline src/elbo.jl:157]: not sure if we can find a better way of formulating it in code, but I did make a double-take and got a bit confused by the type parametrisation making it seem like it was an exponential link - not invlink - function.. what do the rest of you think ?

not helped by link functions being invertible ... I'm wondering if it'd be more helpful to somehow encode input/output domains (e.g. "ExpLink() isa LinkFunction{with input domain (-infty, +infty) and output domain (0, infty)"

File: src/svgp.jl [hunk 21]
```diff
@@ -0,0 +1,68 @@
+struct SVGP end
+
+raw"""
+    approx_posterior(::SVGP, fz::FiniteGP, q::MvNormal)
```
Comment (st--) [inline src/svgp.jl:4]: Maybe this is better turned into its own issue: Have we hardcoded that "inducing variable" is only u = f(Z)? How would we support generic inducing variables (Fourier features, spherical harmonics, or even conditioning on gradients)?

File: src/svgp.jl [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+```math
+q(f^*) = \int p(f | u) q(u) du
+```
+which can be found in closed form.
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+function approx_posterior(::SVGP, fz::FiniteGP, q::AbstractMvNormal)
```
Comment (st--) [inline src/svgp.jl:20]: Is it intentional to not extend `AbstractGPs.approx_posterior`?

File: examples/regression.jl (hunk 17 above)
Comment (rossviljoen) [reply examples/regression.jl:1]: They're not set up as notebooks or anything yet - they're basically just scripts at the moment. I was going to set up proper example notebooks with their own dependencies etc later.

File: src/svgp.jl (hunk 22 above)
Comment (rossviljoen) [reply src/svgp.jl:20]: No, I thought I'd fixed that.. thanks!

File: src/elbo.jl (hunk 19 above)
Comment (rossviljoen) [reply src/elbo.jl:1]: I'm not sure what I meant by that....

File: src/elbo.jl (hunk 13 above)
Comment (rossviljoen) [reply src/elbo.jl:39]: Yeah, it's just the first element of Σy

File: src/elbo.jl (hunk 20 above)
Comment (st--) [reply src/elbo.jl:157]: Well, I see it as an unfortunate legacy from statistics, it's just that it's very well defined that "f = link(expectation(y))" so when we think about transforming f to get the mean parameter e.g. for Poisson likelihood, then lambda = invlink(f). And `PoissonLikelihood{ExpLink}` then makes it seem like it might be a "log invlink" (not that that makes any sense)...

File: examples/regression.jl (hunk 17 above)
Comment (st--) [reply examples/regression.jl:1]: Ah, I was just hoping for recommended instructions to see a rendered version : )

File: src/elbo.jl (hunk 13 above)
Comment (st--) [reply src/elbo.jl:39]: is that because length(Σy) == 1?

File: src/elbo.jl (hunk 18 above)
Comment (willtebbutt) [reply src/elbo.jl]: It's not clear to me how this convention would compromise performance. Perhaps we should discuss elsewhere?

File: src/elbo.jl (hunk 13 above)
Comment (willtebbutt) [reply src/elbo.jl:39]: I think it's because we're assuming it to be constant? (which is something that we should enforce via dispatch)

File: src/elbo.jl [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+    fz::FiniteGP,
+    q::AbstractMvNormal;
+    n_data=length(y),
+    method=Default()
+)
+    return _elbo(method, fx, y, fz, q, GaussianLikelihood(fx.Σy[1]), n_data)
+end
+
+function AbstractGPs.elbo(::FiniteGP, ::AbstractVector, ::FiniteGP, ::AbstractMvNormal; kwargs...) where T<:FiniteGP
+    return error("The observation noise fx.Σy may not be homoscedastic.\n To avoid this error, construct fx using: f = GP(kernel); fx = f(x, σ²)")
```
Comment (willtebbutt) [inline src/elbo.jl:57]: Do you mean heteroscedastic here?
```suggestion
    return error("The observation noise fx.Σy may not be heteroscedastic.\n To avoid this error, construct fx using: f = GP(kernel); fx = f(x, σ²)")
```

**Review by willtebbutt — COMMENTED** [review]
Just a minor typo I think?

File: src/elbo.jl (hunk 23 above)
Comment (rossviljoen) [reply src/elbo.jl:57]: I meant it as in "the noise you supplied might not be homoscedastic, you should fix it", but yeah the phrasing's terrible - what about `The observation noise fx.Σy must be homoscedastic...`?

Comment (willtebbutt) [reply src/elbo.jl]: Ohhh okay. Yeah, that sounds good to me.

File: src/elbo.jl (hunk 12 above)
Comment (rossviljoen) [reply src/elbo.jl]: I've changed `elbo` to dispatch on the type of `fx.Σy` - I'm not 100% sure about the way I've done it and the style though, so feedback would be v appreciated.

File: src/elbo.jl (hunk 13 above)
Comment (rossviljoen) [reply src/elbo.jl:39]: Changed it do dispatch on the type of `Σy` - see my comment above (to keep discussion in one place)

File: src/svgp.jl (hunk 21 above)
Comment (rossviljoen) [reply src/svgp.jl:4]: Opened #11

File: examples/classification.jl [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+# %%
+# First, create the GP kernel from given parameters k
+function make_kernel(k)
+    return softplus(k[1]) * (SqExponentialKernel() ∘ ScaleTransform(softplus(k[2])))
+end
+
+k = [10, 0.1]
+
+kernel = make_kernel(k)
+f = LatentGP(GP(kernel), BernoulliLikelihood(), 0.1)
```
Comment (st--) [inline examples/classification.jl:36]: Maybe something that deserves better documentation (or an outer constructor with a sensible default value), but LatentGP's `Σy` field (can we find a more sensible name!?) is effectively just the jitter for numeric stability, so should be more something like 1e-6 or so...

File: examples/classification.jl [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+    seriescolor="red",
+    linealpha=0.2,
+    label=""
+)
+scatter!(plt, x, y; seriescolor="blue", label="Data points")
+
+# %%
+# Plot the same samples, but pushed through a logistic sigmoid to constrain
+# them in (0, 1).
+prior_y_samples = mean.(f.lik.(prior_f_samples))
```
Comment (st--) [inline examples/classification.jl:57]: Might be instructive to show the full data generating process, including sampling y based on the likelihood, in addition to plotting the result of passing the latent f sample through the invlink function - happy to leave this for future PR though

File: examples/classification.jl [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+    A # variational covariance
+    z # inducing points
+end
+
+@Flux.functor SVGPModel (k, m, A,) # Don't train the inducing inputs
+
+lik = BernoulliLikelihood()
+function (m::SVGPModel)(x)
+    kernel = make_kernel(m.k)
+    f = LatentGP(GP(kernel), BernoulliLikelihood(), 0.1)
```
Comment (st--) [inline examples/classification.jl:85]:
```suggestion
    f = LatentGP(GP(kernel), BernoulliLikelihood(), 1e-6)
```

File: examples/regression.jl [hunk 27]
```diff
@@ -0,0 +1,178 @@
...
+
+# Create the 'model' from the parameters - i.e. return the FiniteGP at inputs x,
+# the FiniteGP at inducing inputs z and the variational posterior over inducing
+# points - q(u).
+function (m::SVGPModel)(x)
+    kernel = make_kernel(m.k)
+    f = GP(kernel)
+    q = MvNormal(m.m, m.A'm.A)
+    fx = f(x, lik_noise)
+    fu = f(m.z, jitter)
```
Comment (st--) [inline examples/regression.jl:57]: nice - giving the constants proper names really helps making it clearer what's going on!

File: src/elbo.jl [hunk 28]
```diff
@@ -0,0 +1,247 @@
...
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+function AbstractGPs.elbo(
+    fx::FiniteGP{<:AbstractGP, <:AbstractVector, <:Diagonal{<:Real, <:Fill}},
+    y::AbstractVector{<:Real},
+    fz::FiniteGP,
+    q::AbstractMvNormal;
```
Comment (st--) [inline src/elbo.jl:49]: does this count as type piracy?

maybe we should leave this for separate discussion in an issue / future refactoring, but I'm wondering about whether it'd be helpful to have an object that contains everything needed for p(f | u) [for u=f(Z) that might be `fz` already, but might need to be something different for interdomain inducing variables] as well as q(u) = `q`]

File: src/elbo.jl [hunk 29]
```diff
@@ -0,0 +1,247 @@
...
+    fx::FiniteGP,
+    y::AbstractVector,
+    fz::FiniteGP,
+    q::AbstractMvNormal,
+    lik::ScalarLikelihood,
+    n_data::Integer
+)
+    post = approx_posterior(SVGP(), fz, q)
+    f_mean, f_var = mean_and_var(post, fx.x)
+    variational_exp = expected_loglik(method, y, f_mean, f_var, lik)
```
Comment (st--) [inline src/elbo.jl:88]: The only reason `_elbo` needs `fx` is to access `x`, right? what do you think of passing `x` as an explicit argument and writing
```suggestion
    variational_exp = expected_loglik(method, y, post(x), lik)
```
- how would that work ?

File: src/elbo.jl (hunk 20 above)
Comment (st--) [reply src/elbo.jl:160]: Could we just call it `Exp` ? It would be even nicer if we could just dispatch on `exp` itself... :thinking:

File: src/elbo.jl [hunk 30]
```diff
@@ -0,0 +1,247 @@
...
+
+    kl_term = StatsBase.kldivergence(q, fz)
```
Comment (st--) [inline src/elbo.jl:90]:
```suggestion
    kl_term = kldivergence(q, fz)
```
should work just the same

File: src/svgp.jl [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+```
+which can be found in closed form.
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+function AbstractGPs.approx_posterior(::SVGP, fz::FiniteGP, q::AbstractMvNormal)
+    m, A = mean(q), cholesky(cov(q))
+    Kuu = cholesky(Symmetric(cov(fz)))
```
Comment (st--) [inline src/svgp.jl:22]: does the cov(q) need a `Symmetric` as well? what's the difference to using AbstractGPs's `_symmetric` (that you also import)?

File: test/test_utils.jl [hunk 32]
```diff
@@ -0,0 +1,15 @@
+# Create a default kernel from two parameters k[1] and k[2]
+make_kernel(k) = softplus(k[1]) * (SqExponentialKernel() ∘ ScaleTransform(softplus(k[2])))
+
+# Computes the optimal closed form solution for the variational posterior
+# q(u) (e.g. # https://krasserm.github.io/2020/12/12/gaussian-processes-sparse/
+# equations (11) & (12)). Assumes a ZeroMean function.
+function exact_variational_posterior(fu, fx, y)
```
Comment (st--) [inline test/test_utils.jl:7]: Could we encode the assumption in code, e.g.
```suggestion
# equations (11) & (12)).
function exact_variational_posterior(fu, fx, y)
    @assert fu.f.mean isa ZeroMean
```
?

Comment (theogf) [reply test/test_utils.jl:7]: `@assert` is generally not a good (check its docs). It's better to have something like `fu.f.mean isa ZeroMean || error("Nope")`

File: src/elbo.jl (hunk 28 above)
Comment (st--) [reply src/elbo.jl:49]: @willtebbutt @theogf

File: examples/classification.jl (hunk 24 above)
Comment (rossviljoen) [reply examples/classification.jl:36]: Looks like it needs at least `1e-4` jitter to work well which is interesting - I'll completely redo the examples in another PR after this one is merged though, so let's discuss it there?

File: src/svgp.jl (hunk 31 above)
Comment (rossviljoen) [reply src/svgp.jl:22]: I removed the import of `_symmetric` because I don't think it's really necessary here - and I suppose it's better not to depend on internals of AbstractGPs.

I suppose it should really be `cholesky(Symmetric(cov(q)))`, but that would break the zero overhead cholesky for `MvNormal` from [here](https://github.com/JuliaStats/Distributions.jl/pull/1373) no?

File: src/elbo.jl (hunk 20 above)
Comment (rossviljoen) [reply src/elbo.jl:160]: Should we move this to an issue in GPLikelihoods?

File: src/svgp.jl (hunk 31 above)
Comment (rossviljoen) [reply src/svgp.jl]: I've changed it to dispatch on the type of `q` - which should solve this

File: src/elbo.jl (hunk 3 above)
Comment (rossviljoen) [reply src/elbo.jl]: Have now switched to using `KLDivergences.jl` for this

File: src/elbo.jl (hunk 15 above)
Comment (rossviljoen) [reply src/elbo.jl]: Changed this to use `AbstractVector{<:Normal}` instead

**Comment by rossviljoen** [comment]
@st-- @willtebbutt @theogf
I'd like to merge this PR soon because it's getting pretty big - could you give it another quick review when you get the chance?

File: src/elbo.jl (hunk 28 above)
Comment (willtebbutt) [reply src/elbo.jl:49]: I think refactoring this a little to put more information in a single object would be a good idea, but I'm happy for this to go in as-is for now.

File: src/elbo.jl [hunk 33]
```diff
@@ -0,0 +1,235 @@
...
+    return sum((y .* f_μ) - exp.(f_μ .+ (var.(q_f) / 2)) - loggamma.(y .+ 1))
+end
+
+# The closed form solution for an Exponential likelihood with an exponential inverse link function
+function expected_loglik(
+    ::Analytic,
+    y::AbstractVector{<:Real},
+    q_f::AbstractVector{<:Normal},
+    ::ExponentialLikelihood{ExpLink}
+    )
```
Comment (willtebbutt) [inline src/elbo.jl:168]:
```suggestion
)
```

File: src/elbo.jl [hunk 34]
```diff
@@ -0,0 +1,235 @@
...
+    return sum(-f_μ - y .* exp.((var.(q_f) / 2) .- f_μ))
+end
+
+# The closed form solution for a Gamma likelihood with an exponential inverse link function
+function expected_loglik(
+    ::Analytic,
+    y::AbstractVector{<:Real},
+    q_f::AbstractVector{<:Normal},
+    lik::GammaLikelihood{<:Any, ExpLink}
+    )
```
Comment (willtebbutt) [inline src/elbo.jl:179]:
```suggestion
)
```

File: src/elbo.jl [hunk 35]
```diff
@@ -0,0 +1,235 @@
...
+    f_μ = mean.(q_f)
+    return sum((lik.α - 1) * log.(y) .- y .* exp.((var.(q_f) / 2) .- f_μ)
+               .- lik.α * f_μ .- loggamma(lik.α))
```
Comment (willtebbutt) [inline src/elbo.jl:182]:
```suggestion
           .- lik.α * f_μ .- loggamma(lik.α))
```

File: src/svgp.jl [hunk 36]
```diff
@@ -0,0 +1,71 @@
...
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+function AbstractGPs.approx_posterior(::SVGP, fz::FiniteGP, q::AbstractMvNormal)
+    m, A = mean(q), _chol_cov(q)
+    Kuu = _chol_cov(fz)
+    B = Kuu.L \ A.L
+    α=Kuu \ (m - mean(fz))
```
Comment (willtebbutt) [inline src/svgp.jl:24]:
```suggestion
    α = Kuu \ (m - mean(fz))
```

**Review by willtebbutt — APPROVED** [review]
I'm happy with this modulo some style stuff. Please feel free to merge when you're happy.
